### ISSUE TITLE
Strict naming convention for function objects

### ISSUE THREAD

**Issue by jaeandersson** [casadi/casadi#1588]
We should enforce a strict naming convention for function objects, i.e.
```python
f = SXFunction(name, ...)
```
should throw an error whenever "name" is not a suitable name.

I think we should allow a-z, A-Z, 0-9, underscore, dash. That will make "name + .c" a valid file name.

**Comment by ghorn** [comment]
Is the motivation for this making C code generation easier?

**Comment by jaeandersson** [comment]
Yes, that's one of the motivations. I also think it will make the printout to screen much more readable. `my_nice_function(x,y)` instead of `My Ugly Function(x,y)` or `(My Ugly Function)(x,y)` etc.

**Comment by ghorn** [comment]
One counterargument: If I have an `ExternalFunction` I might want to call it `my_ode.so` to clarify that it is loaded. I think the user should be able to choose. If someone wants to do `SX.sym('x[0]/y[0]')` for some crazy reason I don't think we should force them to do differently, unless it's specifically required for some implementation detail.

**Comment by jaeandersson** [comment]
What if you name your symbols `SX.sym(' */ hehe, now i have escaped the comment section and can do some evil things /* ')?

**Comment by jaeandersson** [comment]
I think it's not too much to ask for that the function be named `my_ode_so` instead of `my_ode.so`. I think the key is to have clear rules and that if you follow these rules, you shouldn't need to worry that you won't be able to do e.g. codegen.

Anyway, your `ExternalFunction` argument is not good. ExternalFunction supports codegen, i.e. you can embed an ExternalFunction instance into a symbolic expression and then do codegen. The function will be generated into an externally defined symbol, resolved at compile time. But that only works if the name is also a valid symbol name.

E.g.:
```python
from casadi import *
from os import system
x = MX.sym('x')
f = MXFunction('f', [x], [sin(x)])
f.generate()
system("gcc -fPIC -shared -O3 f.c -o f.so")
f2 = ExternalFunction('f')
[y] = f2([cos(x)])
f3 = MXFunction('f3', [x],[y])
f3.generate() # OUCH, weird symbol names!
```

**Comment by jaeandersson** [comment]
Should have said,  "resolved at link time", but you get the point.

**Comment by ghorn** [comment]
Why not sanitize names only when doing codegen?

**Comment by jaeandersson** [comment]
I see codegen not only as a way to speed up calculations, but as a printout of the virtual machine code. So, if virtual machine name deviates from the codegen'd name, you can't e.g. search for a particular symbol and the c code becomes less readable. If you do profiling, it becomes harder to map the output you get to the names you chose for the function objects. Then there is the question of printing expressions. In a printout of the form `sin(my_f.so(x, cos(y)))`, the dot will look like some kind of operator.

In short, it simply makes our lives easier if name is always sane.

**Comment by ghorn** [comment]
OK I'm sold

**Comment by jaeandersson** [comment]
Good. This is a feature that needs gentle deprecation. Warning message in CasADi 2.5, removed post-2.5.

**Comment by jaeandersson** [comment]
I propose to allow:
 * Upper and lower case letters: a-zA-Z
 * Numbers 0-9, but never as first character
 * Underscore, but never as first character and never next to another underscore
 * Letters in the Swedish alphabet: å, ä, ö, Å, Ä, Ö 
 * At most 255 characters

Ping @ghorn @jgillis

**Comment by jaeandersson** [comment]
Ping @toivoh

**Comment by toivoh** [comment]
If you do this, please provide a function that can be used to sanitize names so that they will follow the rules. In some cases I would want to create a function with a name that comes from the user, e.g. the name of a Modelica function (which I'm quite sure allows some function names that don't satisfy the above).

Also,
* It seems to be a good idea to that all allowed names are valid identifiers in C, but that precludes åäö, right? Also seems strange to allow those but not ü, greek letters, etc. etc.
* Why the restrictions on underscores? To reserve some name space for generated names?

**Comment by jaeandersson** [comment]
> It seems to be a good idea to that all allowed names are valid identifiers in C, but that precludes åäö, right? Also seems strange to allow those but not ü, greek letters, etc. etc.

It was actually intended as a joke.. to see if someone would react... Of course we should't allow those characters;)

> Why the restrictions on underscores? To reserve some name space for generated names?

See http://stackoverflow.com/questions/228783/what-are-the-rules-about-using-an-underscore-in-a-c-identifier. Leading underscore or double underscore are "reserved" in C++.

**Comment by jaeandersson** [comment]
> If you do this, please provide a function that can be used to sanitize names so that they will follow the rules. In some cases I would want to create a function with a name that comes from the user, e.g. the name of a Modelica function (which I'm quite sure allows some function names that don't satisfy the above).

I'm not sure if that's a good idea. Names coming directly from the user could be way too long and unreadable anyway. I think it's better idea for the user to maintain a map to some index:
 * `v433`
 * `v1354`
 * `v143423`
Alternatively a more compact name plus an index:
 * `v433_flowrate`
 * `v1354_z0`
 * `v143423_u1`
Then you have to keep a separate table for the full name corresponding to each index.

**Comment by toivoh** [comment]
Well if you don't write such a function then please make sure that the
rules will stay the same, because I might have to write one.

**Comment by jaeandersson** [comment]
But how will you handle long function names from Modelica?

**Comment by toivoh** [comment]
It would be a reasonable effort kind of thing (no guarantees against name collisions). To long names would probably just be truncated.

**Comment by jaeandersson** [comment]
Yeah, but if you cannot guarantee against name collisions, you need an index anyway, right? So you'd name your with the index and then include as many characters as you can?

**Comment by jaeandersson** [comment]
I think the key is if you want to make the a one-to-one map from a generalized character set to a more restricted character set. I don't think it makes sense to go down that route.

**Comment by jaeandersson** [comment]
Note that there is already a function for making something a valid name. It's (currently) called `Function::sanitizeName(const std::string& name)`.

**Comment by toivoh** [comment]
Aha, thanks! That was what I was asking for. Yeah, I agree that a
one-to-mapping is too ambitious.

**Comment by jaeandersson** [comment]
Postponed till 3.1
